### ISSUE TITLE
spread.pb: python.reflect.qual() gets a string which it cannot handle

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7676]
|<img alt="wolfgang61's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| wolfgang61 reported|
|-|-|
|Trac ID|trac#7676|
|Type|release blocker: regression|
|Created|2014-10-11 18:15:13Z|
|Branch|https://github.com/twisted/twisted/tree/log-exception-catching-7676-1|

This is the source from trunk. 

Attached is a test script reproducing the bug:
```
  File "/home/wr/vpn/kajongg/src/twisted/python/reflect.py", line 365, in qual
    return clazz.__module__ + '.' + clazz.__name__
exceptions.AttributeError: 'str' object has no attribute '__module__'

```

. I would like to convert that into a test case but I am not sure how to use StringTransport in this context. What I tried is in the script as dead code.

The trivial fix would be to make qual(clazz) just return str(clazz) if clazz has no __module__ attribute. But then I am not sure: Maybe the bug is really in the caller?

Attachments:

* [ticket7676.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/06e/06e80210c4d05684c666906d9e2ea73355087d0b/6f0242b5486b2349dd8c0a93c866936d4a7ac86e.py) (1473 bytes) - added by wolfgang61 on 2014-10-11 18:16:36Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__7676 7676
type__release_blocker__regression release blocker: regression
reporter__wolfgang61 wolfgang61
priority__highest highest
milestone__New_Logging_System New Logging System
branch__branches_log_exception_catching_7676_1 branches/log-exception-catching-7676-1
branch_author__ 
status__closed closed
resolution__fixed fixed
component__logger logger
keywords__logging_ logging,
time__1413051313221034 1413051313221034
changetime__1481166270879945 1481166270879945
version__None None
owner__wsanchez wsanchez

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun set owner to wolfgang61|
|-|-|

I don't see the described error.  When I run the attachment, I see:

```
Peer will receive following PB traceback:
Unhandled Error
Traceback (most recent call last):
  File "/home/exarkun/.local/lib/python2.7/site-packages/twisted/spread/banana.py", line 153, in gotItem
    self.callExpressionReceived(item)
  File "/home/exarkun/.local/lib/python2.7/site-packages/twisted/spread/banana.py", line 116, in callExpressionReceived
    self.expressionReceived(obj)
  File "/home/exarkun/.local/lib/python2.7/site-packages/twisted/spread/pb.py", line 565, in expressionReceived
    method(*sexp[1:])
  File "/home/exarkun/.local/lib/python2.7/site-packages/twisted/spread/pb.py", line 877, in proto_message
    self._recvMessage(self.localObjectForID, requestID, objectID, message, answerRequired, netArgs, netKw)
--- <exception caught here> ---
  File "/home/exarkun/.local/lib/python2.7/site-packages/twisted/spread/pb.py", line 891, in _recvMessage
    netResult = object.remoteMessageReceived(self, message, netArgs, netKw)
  File "/home/exarkun/.local/lib/python2.7/site-packages/twisted/spread/flavors.py", line 112, in remoteMessageReceived
    raise NoSuchMethod("No such method: remote_%s" % (message,))
twisted.spread.flavors.NoSuchMethod: No such method: remote_clientMethod
Unhandled error in Deferred:
Unhandled Error
Traceback from remote host -- Traceback unavailable
twisted.spread.flavors.NoSuchMethod: No such method: remote_clientMethod
```

**Comment by twisted-trac** [comment]
|<img alt="wolfgang61's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|wolfgang61 set owner to @exarkun|
|-|-|

Did you really use latest trunk? With 14.0.0, I get your behaviour, with trunk I get mine.

git bisect tells me this is a regression from svn r42936:

FileLogObserver now appends a traceback when writing an event that contains the log_event key.
    HOWTO now notes log_failure key and explains the log.failure method.

But it seems there are more situations where reflect.qual() fails.

I now retested my full application against twisted-14.0.0, and I still get the same error. 
Problem is that I reduced my app step by step down to the posted test script, but always with twisted-trunk. So with twisted-14.0.0, I would have gotten yet another test script taking a different path to that bug. But I really do not want to repeat that exercise, it just took too much time.

With twisted-13.0.0, my app does not trigger this bug. I can git-bisect to find the relevant commit between 13.0.0 and 14.0.0 if you want.

Yet more: The commits between 14.0.0 and trunk that I tested while bisecting showed more differences in behaviour, some produced more error output, some less. Again if you want to know I can find relevant commits and post the differences.

So since all this looks a bit shaky, I would not be surprised if there were more paths to triggering this bug.

So I really believe the right fix is to make qual() more resistant against argument values without the ```__module__``` attribute.

BTW: With PY2, my app only triggers this after inserting an ```assert 0``` in a nice place. With PY3, it always triggers this (needs some more porting). Random injection of ```assert 0``` might be an interesting idea for automated test loops. Comparing the behaviour between twisted releases.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

> Did you really use latest trunk? With 14.0.0, I get your behaviour, with trunk I get mine.

I didn't because I didn't realize this only affected trunk.  Testing against trunk I get this behavior:

```
Traceback (most recent call last):
  File "/home/exarkun/Projects/Twisted/trunk/twisted/spread/pb.py", line 563, in expressionReceived
    method(*sexp[1:])
  File "/home/exarkun/Projects/Twisted/trunk/twisted/spread/pb.py", line 975, in proto_error
    d.errback(self.unserialize(fail))
  File "/home/exarkun/Projects/Twisted/trunk/twisted/internet/defer.py", line 426, in errback
    self._startRunCallbacks(fail)
  File "/home/exarkun/Projects/Twisted/trunk/twisted/internet/defer.py", line 493, in _startRunCallbacks
    self._runCallbacks()
--- <exception caught here> ---
  File "/home/exarkun/Projects/Twisted/trunk/twisted/internet/defer.py", line 580, in _runCallbacks
    current.result = callback(current.result, *args, **kw)
  File "/home/exarkun/Projects/Twisted/trunk/twisted/spread/pb.py", line 944, in _sendFailureOrError
    self._sendFailure(fail, requestID)
  File "/home/exarkun/Projects/Twisted/trunk/twisted/spread/pb.py", line 952, in _sendFailure
    log.err(fail)
  File "/home/exarkun/Projects/Twisted/trunk/twisted/python/log.py", line 133, in err
    msg(failure=_stuff, why=_why, isError=1, **kw)
  File "/home/exarkun/Projects/Twisted/trunk/twisted/python/threadable.py", line 53, in sync
    return function(self, *args, **kwargs)
  File "/home/exarkun/Projects/Twisted/trunk/twisted/python/log.py", line 281, in msg
    _publishNew(self._publishPublisher, actualEventDict, textFromEventDict)
  File "/home/exarkun/Projects/Twisted/trunk/twisted/python/logger/_legacy.py", line 222, in publishToNewObserver
    text = textFromEventDict(eventDict)
  File "/home/exarkun/Projects/Twisted/trunk/twisted/python/log.py", line 422, in textFromEventDict
    text = (why + '\n' + eventDict['failure'].getTraceback())
  File "/home/exarkun/Projects/Twisted/trunk/twisted/python/failure.py", line 531, in getTraceback
    self.printTraceback(file=io, elideFrameworkCode=elideFrameworkCode, detail=detail)
  File "/home/exarkun/Projects/Twisted/trunk/twisted/python/failure.py", line 592, in printTraceback
    w("%s: %s\n" % (reflect.qual(self.type),
  File "/home/exarkun/Projects/Twisted/trunk/twisted/python/reflect.py", line 364, in qual
    return clazz.__module__ + '.' + clazz.__name__
exceptions.AttributeError: 'str' object has no attribute '__module__'
```

which sounds like what you encountered.

> The trivial fix would be to make `qual(clazz)` just return `str(clazz)` if clazz has no module attribute. But then I am not sure:

Indeed, all that would do in this case is tell you misformat the resulting traceback to claim that `__builtin__.str` was raised.

> Maybe the bug is really in the caller?

It seems like.  Actually, the bug maybe in the application code that triggers this:

```python
    def failed(self, f):
        raise f
```

This is bad.  This is not how you propagate a `Failure`.  Of course, Twisted is at least partially to blame since unfortunately there is code in Twisted to support this usage.  It's not code that works entirely reliably.  I'm not sure why this _ever_ behaved as you observed it previously behaved, but I think it only did so by accident.

In this example, because you're using PB, `f` is actually a `CopiedFailure` - not a `Failure`.  `CopiedFailure` has special code to support formatting a traceback that was produced remotely.  Part of this is dealing with the (horrible) fact that `CopiedFailure.type` is the FQPN of the exception type, not the exception type itself.

However, `raise aCopiedFailure` changes the type to `Failure` (because the raised `CopiedFailure` is caught by the implementation of `Deferred` and treated as any normal exception from the callback method; it constructs a new `Failure` to represent this state - and that new `Failure` notices that what you've actually done is raise another `Failure` (or a `CopiedFailure`, in this case - a subclass of `Failure`) so it grabs all of the state off of that other object, making it appear as though identity has been preserved).

`Failure.getTraceback` doesn't know how to deal with `type` being a string so everything blows up.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to @wsanchez|
|-|-|

Here's a shorter example that demonstrates essentially the same logging bug that I think is the real regression being encountered here:

```
from twisted.spread.pb import CopiedFailure
from twisted.python.log import err

try:
        1 / 0
except:
        err(CopiedFailure())
```

The symptom is slightly different (it complains about a missing `traceback` attribute) but the underlying cause is the same: `twisted.python.log.err` now lets log observers raise exceptions up to the caller of `err` whereas previously it protected the caller from such exceptions.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph commented|
|-|-|

*klaxon noises*

**Comment by twisted-trac** [comment]
|<img alt="wolfgang61's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|wolfgang61 commented|
|-|-|

Replying to [exarkun](#note_3):
> > The trivial fix would be to make `qual(clazz)` just return `str(clazz)` if clazz has no module attribute. But then I am not sure:
> 
> Indeed, all that would do in this case is tell you misformat the resulting traceback to claim that `__builtin__.str` was raised.

No, it would return "twisted.spread.flavors.NoSuchMethod", using this version of qual():
```
def qual(clazz):
    """
    Return full import path of a class.
    """
    if getattr(clazz, '__module__', None) is None:
        print('**** qual returns %s' % clazz)
        return str(clazz)
    return clazz.__module__ + '.' + clazz.__name__
```

**Comment by twisted-trac** [comment]
|<img alt="wolfgang61's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|wolfgang61 commented|
|-|-|

Replying to [exarkun](#note_3):
> > Maybe the bug is really in the caller?
> 
> It seems like.  Actually, the bug maybe in the application code that triggers this:
> 
> ```
> #!python
>     def failed(self, f):
>         raise f
> ```
> 
> This is bad.  This is not how you propagate a `Failure`.  Of course, Twisted is at least partially to blame since unfortunately there is code in Twisted to support this usage.  It's not code that works entirely reliably.  I'm not sure why this _ever_ behaved as you observed it previously behaved, but I think it only did so by accident.

With "in the caller" I only meant the part of twisted calling this, not the application. I do not care if this use of twisted is legal or not, because it does not matter: Twisted should never show an internal bug instead of the external one. At least I hope so.

So it does not matter what the application does. Anyway, here comes a clarification: My full application did not do this. This is just the smallest example I found which triggers the bug. The original problem was more complex: It calls into Qt4 which calls a method rewritten in Python which triggers an assertion, but only when running with PY3. Because that part of my application does not yet support PY3.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

> With "in the caller" I only meant the part of twisted calling this, not the application. I do not care if this use of twisted is legal or not, because it does not matter: Twisted should never show an internal bug instead of the external one. At least I hope so.

I know.  However, my point is that that application code is _also_ broken.  We'll fix the logging regression here, but that application code is broken on Twisted 14.0 and on trunk@HEAD.

> No, it would return "twisted.spread.flavors.NoSuchMethod", using this version of qual():

That's not a great change to `qual`.  I don't think we'll pursue it.  Fortunately it seems clear what the misbehavior is in the logging system now, so we don't need to.

**Comment by twisted-trac** [comment]
|<img alt="wolfgang61's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|wolfgang61 commented|
|-|-|

I suppose this is the same bug:
```
  File "/home/wr/vpn/kajongg/src/twisted/spread/pb.py", line 586, in expressionReceived
    method(*sexp[1:])
  File "/home/wr/vpn/kajongg/src/twisted/spread/pb.py", line 1007, in proto_error
    d.errback(self.unserialize(fail))
  File "/home/wr/vpn/kajongg/src/twisted/internet/defer.py", line 426, in errback
    self._startRunCallbacks(fail)
  File "/home/wr/vpn/kajongg/src/twisted/internet/defer.py", line 493, in _startRunCallbacks
    self._runCallbacks()
  File "/home/wr/vpn/kajongg/src/twisted/internet/defer.py", line 625, in _runCallbacks
    current.result.cleanFailure()
  File "/home/wr/vpn/kajongg/src/twisted/python/failure.py", line 499, in cleanFailure
    self.value.__traceback__ = None
builtins.AttributeError: 'str' object has no attribute '__traceback__'
```

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)|@wsanchez set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_13"></a>|@wsanchez commented|
|-|-|

This seems to fix the problem:
```
Index: twisted/python/logger/_legacy.py
===================================================================
--- twisted/python/logger/_legacy.py	(revision 43439)
+++ twisted/python/logger/_legacy.py	(working copy)
@@ -219,7 +219,10 @@
     """
 
     if "log_format" not in eventDict:
-        text = textFromEventDict(eventDict)
+        try:
+            text = textFromEventDict(eventDict)
+        except:
+            text = None
         if text is not None:
             eventDict["log_text"] = text
             eventDict["log_format"] = "{log_text}"
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

Perhaps `text` could be something slightly more descriptive than `None` when an exception is handled in this code?

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_15"></a>|@wsanchez commented|
|-|-|

exarkun: I'd agree, but in this case I'm trying to preserve the existing behavior of the old logging system: when it's unable to render text, it renders nothing.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_16"></a>|@wsanchez commented|
|-|-|

Need to merge forward, as there are now tests for C{publishToNewObserver} which didn't exist before and we want to add to.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_17"></a>|@wsanchez commented|
|-|-|

r43694 adds the above fix plus a test.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_19"></a>|@wsanchez commented|
|-|-|

builedbot done.  twistedchecker and pydoctor errors exist on logging branch and are, I think, false positives/bugs.

Ready for review.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)|@wsanchez removed owner<br>@wsanchez set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_21"></a>|@glyph set owner to @wsanchez|
|-|-|

Thanks for following up on this, wsanchez.

I agree about pydoctor and twistedchecker.  I think they're real issues in the larger branch, but the branch-of-a-branch workflow here is confusing the buildbots.

I have a couple of concerns though:

  1. The output isn't exactly the same.  On trunk, from the original example, I get:
```
Peer will receive following PB traceback:
Unhandled Error
Traceback (most recent call last):
  File ".../twisted/spread/banana.py", line 167, in gotItem
    self.callExpressionReceived(item)
  File ".../twisted/spread/banana.py", line 130, in callExpressionReceived
    self.expressionReceived(obj)
  File ".../twisted/spread/pb.py", line 563, in expressionReceived
    method(*sexp[1:])
  File ".../twisted/spread/pb.py", line 875, in proto_message
    self._recvMessage(self.localObjectForID, requestID, objectID, message, answerRequired, netArgs, netKw)
--- <exception caught here> ---
  File ".../twisted/spread/pb.py", line 889, in _recvMessage
    netResult = object.remoteMessageReceived(self, message, netArgs, netKw)
  File ".../twisted/spread/flavors.py", line 112, in remoteMessageReceived
    raise NoSuchMethod("No such method: remote_%s" % (message,))
twisted.spread.flavors.NoSuchMethod: No such method: remote_clientMethod
Unhandled error in Deferred:
Unhandled Error
Traceback from remote host -- Traceback unavailable
twisted.spread.flavors.NoSuchMethod: No such method: remote_clientMethod
```
  and on the branch, I get:
```
Peer will receive following PB traceback:
Unhandled Error
Traceback (most recent call last):
  File ".../twisted/spread/banana.py", line 167, in gotItem
    self.callExpressionReceived(item)
  File ".../twisted/spread/banana.py", line 130, in callExpressionReceived
    self.expressionReceived(obj)
  File ".../twisted/spread/pb.py", line 563, in expressionReceived
    method(*sexp[1:])
  File ".../twisted/spread/pb.py", line 875, in proto_message
    self._recvMessage(self.localObjectForID, requestID, objectID, message, answerRequired, netArgs, netKw)
--- <exception caught here> ---
  File ".../twisted/spread/pb.py", line 889, in _recvMessage
    netResult = object.remoteMessageReceived(self, message, netArgs, netKw)
  File ".../twisted/spread/flavors.py", line 112, in remoteMessageReceived
    raise NoSuchMethod("No such method: remote_%s" % (message,))
twisted.spread.flavors.NoSuchMethod: No such method: remote_clientMethod


Unhandled error in Deferred:


Traceback from remote host -- Traceback unavailable
twisted.spread.flavors.NoSuchMethod: No such method: remote_clientMethod
```
  I don't know where those extra newlines are coming from and it bugs me.  Not that they're that bad, but we shouldn't be randomly emitting extra whitespace without knowing why.  This could be a separate ticket though; like I said I don't know where they're coming from, maybe they show up in other scenarios too.
  2. I kinda feel like the exception handling is in the wrong place here.  In the legacy log observer code, the exception was handled by the code calling the observer, not the code calling textForEventDict.  I guess we're now catching the exception as we are _converting_ from an old-style message to a new-style one... which we didn't have to do before, since formatting was a step performed by the observer.
  3. The fact that textFromEventDict is injected as a fake parameter seems a bit weird and unnecessary.  There are values you can give to the real textFromEventDict that will make it behave in this way, and by doing it that way - giving it a value that makes the real implementation blow up - it's clearer that this is a real situation that can happen and not unnecessary belt-and-suspenders gold plating even though textFromEventDict handles all possible errors already.  The comprehensive use of `_safeFormat` makes one think that `textFromEventDict` is already trying really hard not to raise an exception.  (I realize that it has to be passed as a parameter to avoid a circular import anyway, but you could pass the real one in the test.)

Given that we are explicitly trying to turn this into legacy code, I think none of this feedback is optional - except for that first one, it would be good to at least understand _why_ we're getting different results now.  Depending on what you discover there, please either fix it up and submit for review again, or just go ahead and land this and file a new ticket.

Thanks again!

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_22"></a>|@wsanchez commented|
|-|-|

glyph: on #3… I went and changed the tests to use the real `textFromEventDict` (r43715).  But when I got to `test_textFromEventDictRaises`, which is the one I think is the one you are most interested in here, I found that creating an event that would cause `textFromEventDict` to raise an exception requires pretty specific knowledge of the implementation of `textFromEventDict`, which isn't the thing we're testing.  A change to `textFromEventDict` could cause the test to pass when it shouldn't, I mean.

I agree that it's unclear why that test is there, so I should add a note… r43716.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_23"></a>|@wsanchez commented|
|-|-|

On #2 I think you're suggesting that `textFromEventDict` should try harder and catch these exceptions and that `publishToNewObserver` should not try to do this itself.  I'm less certain about the latter, but I consider belts and suspenders more façionable than you do.  :-)

The former is fixed in r43717, and `textFromEventDict` gets units tests that cover that plus the rest of its functionality.

The latter is unwound in r43718.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_24"></a>|@wsanchez commented|
|-|-|

Still trying to figure out what the deal is with [#11338](https://github.com/twisted/twisted/issues/11338).

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)|@wsanchez set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_26"></a>|@glyph commented|
|-|-|

Replying to [wsanchez](#note_23):
> On #2 I think you're suggesting that `textFromEventDict` should try harder and catch these exceptions and that `publishToNewObserver` should not try to do this itself.

Since `textFromEventDict` doesn't document raising an exception, it seems reasonable that it should try extra-hard not to do so; no valid client could be broken by such a change.  I was waffling on this a little bit but i'm glad you made the change.

> I'm less certain about the latter, but I consider belts and suspenders more façionable than you do.  :-)

Belt and suspenders is OK, the problem with open source projects with many contributors is that they see _you_ wearing a belt and suspenders, and they get the idea that they should be wearing full body armor made entirely out of belts held together with suspenders... ;-)

> The former is fixed in r43717, and `textFromEventDict` gets units tests that cover that plus the rest of its functionality.
>
> The latter is unwound in r43718.

Looks good.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_27"></a>|@glyph commented|
|-|-|

Replying to [wsanchez](#note_22):
> glyph: on #3… I went and changed the tests to use the real `textFromEventDict` (r43715).  But when I got to `test_textFromEventDictRaises`, which is the one I think is the one you are most interested in here, I found that creating an event that would cause `textFromEventDict` to raise an exception requires pretty specific knowledge of the implementation of `textFromEventDict`, which isn't the thing we're testing.  A change to `textFromEventDict` could cause the test to pass when it shouldn't, I mean.
> 
> I agree that it's unclear why that test is there, so I should add a note… r43716.

I think what you've got now is fine, but just as a point of order: we could prevent that change to `textFromEventDict` causing a problem by having a sanity check that ensures it raises an exception with a comment explaining that if it the test began to fail there because `textFromEventDict` changed, it could be changed or deleted.

Hopefully that sentence made sense despite its excessive length ;).

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_28"></a>|@wsanchez commented|
|-|-|

Replying to [glyph](#note_27):

> Hopefully that sentence made sense despite its excessive length ;).

It did, but `textFromEventDict` might still raise due to some bug.  Anyway, not relevant any more because r43718 removed this behavior and test.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_29"></a>|@wsanchez commented|
|-|-|

I've opened [#7755](https://github.com/twisted/twisted/issues/7755) to address the remaining issue with the newlines, since it's not the same problem described here, and can be tracked separately.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_30"></a>|@wsanchez set status to `closed`|
|-|-|

(In [43723]) Merge [branches/log-exception-catching-7676-1:](branches/log-exception-catching-7676-1:) spread.pb: python.reflect.qual() gets a string which it cannot handle

Author: wsanchez
Reviewer: glyph
Fixes: [#7676](https://github.com/twisted/twisted/issues/7676)
